core/domain/question_jobs_one_off.py

```diff
@@ -140,3 +142,86 @@ def reduce(key, values):
                 sum(ast.literal_eval(v) for v in values))])
         else:
             yield (key, values)
+
+
+class MissingQuestionMigrationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """This job is used to delete question commit log models for which question
+    models are missing.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [question_models.QuestionCommitLogEntryModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        question = question_services.get_question_by_id(
+            item.question_id, strict=False)
+        if question is None:
+            item.delete()
+            yield ('Question Commit Model deleted', 1)
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, sum(ast.literal_eval(v) for v in values))
+
+
+class RegenerateQuestionCommitAndSnapshotOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job to regenerate question commit log and snapshot models if they
+    are missing."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [question_models.QuestionModel]
+
+    @staticmethod
+    def map(question_model):
+        if question_model.deleted:
+            return
+
+        for version in python_utils.RANGE(1, question_model.version + 1):
+            commit_log_model = (
+                question_models.QuestionCommitLogEntryModel.get_by_id(
+                    'question-%s-%s' % (question_model.id, version)))
+            if commit_log_model is None or commit_log_model.deleted:
```

Comment:
If it has the 'deleted' field, don't we just have to flip it back to false?

Reply:
Makes sense, done.